stackslib/src/chainstate/stacks/boot/signers-voting.clar

```diff
@@ -77,35 +73,88 @@
             )
         (get prepare-cycle-length pox-info)))
 
+;; get the aggregate public key for the given reward cycle (or none)
+(define-read-only (get-approved-aggregate-key (reward-cycle uint))
+    (map-get? aggregate-public-keys reward-cycle)
+)
+
 (define-private (is-in-voting-window (height uint) (reward-cycle uint))
-    (let ((last-cycle (unwrap-panic (contract-call? .signers stackerdb-get-last-set-cycle))))
+    (let ((last-cycle (unwrap-panic (contract-call? .signers get-last-set-cycle))))
         (and (is-eq last-cycle reward-cycle)
             (is-in-prepare-phase height))))
 
+(define-private (sum-weights (signer { signer: principal, weight: uint }) (acc uint))
+    (+ acc (get weight signer))
+)
+
+(define-private (get-total-weight (reward-cycle uint))
+    (match (map-get? cycle-total-weight reward-cycle)
+        total (ok total)
+        (let (
+                (signers (unwrap! (contract-call? .signers get-signers reward-cycle) (err ERR_FAILED_TO_RETRIEVE_SIGNERS)))
+                (total (fold sum-weights signers u0))
+            )
+            (map-set cycle-total-weight reward-cycle total)
+            (ok total)
+        )
+    )
+)
+
+;; Signer vote for the aggregate public key of the next reward cycle
+;;  The vote happens in the prepare phase of the current reward cycle but may be ran more than
+;;  once resulting in different 'rounds.' Each signer vote is based on the weight of stacked
+;;  stx tokens & fetched from the .signers contract. The vote is ran until the consensus 
+;;  threshold of 70% for a specific aggregate public key is reached. 
 (define-public (vote-for-aggregate-public-key (signer-index uint) (key (buff 33)) (round uint))
```

Comment:
I wonder if this needs to take the reward cycle as an option. 

For example Signer N could be registered for the current and the next reward cycle.

Assume Signer N gets assigned the same slot index in both page 0 and page 1 of the stacker db slot. 

For whatever reason, reward cycle page 0 requires a DKG vote. Simultaneously, reward cycle with page 1 is running a DKG vote. How can you distinguish these two votes if you do not include the reward cycle or page number? (where page number is reward cycle modulus 2 )

Reply:
@setzeus has now added this.